### PR TITLE
KAFKA-13950: Fix resource leak in error scenarios

### PR DESCRIPTION
**Pull request by divijvaidya**
# Problem
We are not properly closing Closeable resources in the code base at multiple places especially when we have an exception.

# Change
This code change fixes multiple of these leaks. Note that `Utils.closeQuietly` is used in this code change where author thinks that propagation of original exception is essential.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java [hunk 1]
```diff
@@ -162,79 +163,82 @@ private static FilterResult filterTo(TopicPartition partition, Iterable<MutableR
...
-                    if (filteredBatchSize > batch.sizeInBytes() && filteredBatchSize > maxRecordBatchSize)
-                        log.warn("Record batch from {} with last offset {} exceeded max record batch size {} after cleaning " +
-                                        "(new size is {}). Consumers with version earlier than 0.10.1.0 may need to " +
-                                        "increase their fetch sizes.",
-                                partition, batch.lastOffset(), maxRecordBatchSize, filteredBatchSize);
-
-                    MemoryRecordsBuilder.RecordsInfo info = builder.info();
-                    filterResult.updateRetainedBatchMetadata(info.maxTimestamp, info.shallowOffsetOfMaxTimestamp,
-                            maxOffset, retainedRecords.size(), filteredBatchSize);
+        try (ByteBufferOutputStream bufferOutputStream = new ByteBufferOutputStream(destinationBuffer)) {
```
Comment (divijvaidya) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:166]: Note that there is no logic change here.

We only added the ByteBufferOutputStream into a try-with-resources block and everything got indented by a tab inside that block.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 2]
```diff
@@ -595,57 +595,65 @@ private WorkerTask buildWorkerTask(ClusterConfigState configState,
...
-                                                                connectorClientConfigOverridePolicy, kafkaClusterId);
-            KafkaProducer<byte[], byte[]> producer = new KafkaProducer<>(producerProps);
-            TopicAdmin admin;
-            Map<String, TopicCreationGroup> topicCreationGroups;
-            if (config.topicCreationEnable() && sourceConfig.usesTopicCreation()) {
-                Map<String, Object> adminProps = adminConfigs(id, "connector-adminclient-" + id, config,
-                        sourceConfig, connectorClass, connectorClientConfigOverridePolicy, kafkaClusterId);
-                admin = new TopicAdmin(adminProps);
-                topicCreationGroups = TopicCreationGroup.configuredGroups(sourceConfig);
+        RetryWithToleranceOperator retryWithToleranceOperator = null;
```
Comment (divijvaidya) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Note that this is intentionally not closed on try-with-resources block because the returned WorkerTask from this method owns it's lifecycle and closes it in it's close() method.

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java [hunk 3]
```diff
@@ -809,6 +810,13 @@ private void maybeCloseOldestConnection(long currentTimeNanos) {
      * poll().
      */
     public void clearCompletedReceives() {
+        this.completedReceives.values().forEach(networkReceive -> {
+            try {
+                networkReceive.close();
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/network/Selector.java]: Can we add a test for this please?

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java (hunk 1 above)
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java]: This is a no-op, so I assume it's just a hygiene thing?

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java [hunk 4]
```diff
@@ -567,12 +577,12 @@ private boolean flushNetOutBuffer() throws IOException {
     private AbstractResponse receiveKafkaResponse() throws IOException {
         if (netInBuffer == null)
             netInBuffer = new NetworkReceive(node);
-        NetworkReceive receive = netInBuffer;
+        final NetworkReceive receive = netInBuffer;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java]: Some of these changes seem unrelated.

**Review by showuon — APPROVED** [review]
LGTM! Nice catch and cleanup!

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java (hunk 3 above)
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/network/Selector.java]: I don't have a good idea on how to assert in the tests that the `NetworkReceive` objects are being closed. I have considered the following:
1. Using `GarbageCollectedMemoryPool` which could have hinted towards a memory leak but it doesn't work.
2. Assert that the NetworkReceive.Close() is being called in the flow but we would need to Spy the objects in the test which is difficult to do here.

I am going to try again on Monday to see if I can figure something out but meanwhile, do you have any suggestions here on how can I add a test to validate the leak?

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java (hunk 4 above)
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java]: That's a fair comment. I will revert these changes to keep the PR clean.

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java (hunk 1 above)
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java]: Yep! Reader of the code may not realize that the ByteBufferOutputStream.Close() is a NoOp (without looking at the code) but with the try-resources block, reader won't have to think about resource leaks and also, it's a nice hint to signify the lifecycle of the object (which is within the try block).

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java (hunk 4 above)
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java]: This is fixed

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java (hunk 3 above)
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/network/Selector.java]: ok, so I was able to write a test that existing code fails by adding a method in `GarbageCollectedMemoryPool` which validates that all buffer has been released. But the memory pool in a selector is not used by anything other than tests. So, production won't have a memory leak here.

The ownership of lifecycle of `NetworkReceive` is confusing and not well defined in the current code. Currently, the list of `CompletedNetworkReceive` is cleared on every poll event (without closing the `NetworkReceive`). Before the poll, a potential caller can obtain reference to `NetworkReceive` using the `Selector.completedReceives()` method. If we close the `NetworkReceive`, the reference available with the caller will not be valid. Which begs the question, is the Selector responsible for closing completed `NetworkReceive(s)` or is it the caller of `Selector.completedReceives()`. The test `SelectorTest.testMuteOnOOM` relies on the fact that caller is responsible for closing the `NetworkReceive`. Although, personally, that doesn't look right to me but nevertheless I don't want to have a discussion about that in the scope of this PR. Since MemoryPool backed `NetworkReceive` is not used by any non-test classes, I am reverting my changes to this class from this PR and will open a separate JIRA ticket to have the discussion about lifecycle ownership.

I hope the above makes sense?

**Comment by divijvaidya** [comment]
@ijuma I believe I have addressed all your comments and this is ready for second round of review from you (pending test run). I also want to thank you for your engagement on this so far. Cheers!

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java (hunk 3 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/network/Selector.java]: Yes, that makes sense. I was unsure how we could have such a bad leak that had gone undetected. It makes sense that the Selector cannot be responsible for closing the network receive since it transfers the ownership to the caller of `poll`. Perhaps we can add a comment explaining this?

Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/network/Selector.java]: Thanks. I added the following comment to the `Selector.completedReceives()` in the latest commit.
```
Note that the caller of this method assumes responsibility to close the NetworkReceive resources which may be
backed by a {@link MemoryPool}. In such scenarios (when NetworkReceive uses a {@link MemoryPool}, it is necessary
to close the {@link NetworkReceive} to prevent any memory leaks.
```

The new revision is ready for your review.

**Comment by divijvaidya** [comment]
@ijuma I have addressed your comments. Please review when you get a chance.

**Comment by cadonna** [comment]
@divijvaidya @jsancio The Streams related changes LGTM!

File: clients/src/main/java/org/apache/kafka/common/network/PlaintextChannelBuilder.java [hunk 5]
```diff
@@ -54,12 +54,18 @@ public void configure(Map<String, ?> configs) throws KafkaException {
...
+            transportLayer = buildTransportLayer(key);
+            final PlaintextTransportLayer finalTransportLayer = transportLayer;
+            Supplier<Authenticator> authenticatorCreator = () -> new PlaintextAuthenticator(configs, finalTransportLayer, listenerName);
             return buildChannel(id, transportLayer, authenticatorCreator, maxReceiveSize,
                     memoryPool != null ? memoryPool : MemoryPool.NONE, metadataRegistry);
         } catch (Exception e) {
+            // Ideally these resources are closed by the KafkaChannel but this builder should close the resources instead
+            // if an error occurs due to which KafkaChannel is not created.
+            Utils.closeQuietly(transportLayer, "transport layer for channel Id: " + id);
+            Utils.closeQuietly(metadataRegistry, "metadataRegistry");
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/network/PlaintextChannelBuilder.java]: I looked into this a bit more and it's a bit odd how we're doing clean-up. `buildAndAttach` currently closes the socket channel and cancels the key. It also creates the selector channel metadata registry. Shouldn't that take care of closing the metadata registry since it actually creates it?

With regards to the transport layer, I wonder if we could move the cleanup logic to `KafkaChannel` instead. We could change `KafkaChannel` to take a `TransportLayer` supplier instead (or add a static factory method) and then have it handle the cleanup (instead of doing it in every implementation of channel builder.

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java [hunk 6]
```diff
@@ -342,6 +342,7 @@ public void reauthenticate(ReauthenticationContext reauthenticationContext) thro
         reauthInfo.reauthenticating(apiVersionsResponseFromOriginalAuthentication,
                 reauthenticationContext.reauthenticationBeginNanos());
         NetworkReceive netInBufferFromChannel = reauthenticationContext.networkReceive();
+        resetNetworkIn(); // reset the buffer before new assignment
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java]: Why do we need to do this?

File: core/src/main/java/kafka/server/builders/ReplicaManagerBuilder.java [hunk 7]
```diff
@@ -147,11 +147,14 @@ public ReplicaManagerBuilder setThreadNamePrefix(String threadNamePrefix) {
 
     public ReplicaManager build() {
         if (config == null) config = new KafkaConfig(Collections.emptyMap());
-        if (metrics == null) metrics = new Metrics();
         if (logManager == null) throw new RuntimeException("You must set logManager");
         if (metadataCache == null) throw new RuntimeException("You must set metadataCache");
         if (logDirFailureChannel == null) throw new RuntimeException("You must set logDirFailureChannel");
         if (alterPartitionManager == null) throw new RuntimeException("You must set alterIsrManager");
+        // Initialize metrics in the end just before passing it to ReplicaManager because ReplicaManager closes the
```
Comment (ijuma) [inline core/src/main/java/kafka/server/builders/ReplicaManagerBuilder.java]: "because" -> "to ensure"?

File: clients/src/main/java/org/apache/kafka/common/network/PlaintextChannelBuilder.java (hunk 5 above)
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/network/PlaintextChannelBuilder.java]: > Shouldn't that take care of closing the metadata registry since it actually creates it?

I agree, I missed it earlier. I have moved the responsibility of closing in case of errors to `Selector.buildAndAttachKafkaChannel`. I have also added a unit test that fails before this fix and passes afterwards.

> With regards to the transport layer, I wonder if we could move the cleanup logic to `KafkaChannel` instead

I am afraid that might not be possible because we are handling scenarios where we encounter an error while creating the `KafkaChannel` itself (e.g. a scenario where authenticatorCreator throws an error in the constructor). We can close it uniformly at `ChannelBuilder` but that would require a bit of refactoring involving conversion of `ChannelBuiler` interface into a class. Please let me know if you have any ideas.

File: core/src/main/java/kafka/server/builders/ReplicaManagerBuilder.java (hunk 7 above)
Comment (divijvaidya) [reply core/src/main/java/kafka/server/builders/ReplicaManagerBuilder.java]: Fixed in upcoming commit

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java (hunk 6 above)
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java]: In the very next line, we are overwriting the `netInBuffer` member and hence losing it's earlier value without having closed it properly. Given the sensitivity of the change in network component here, I am removing all changes to this buffer from this PR to unblock it and will file a separate PR to address this.

**Comment by divijvaidya** [comment]
@ijuma what are your thoughts on this one? I would either like to discard it or have it merged since it is pending for a long time! Let me know your thoughts.

**Comment by divijvaidya** [comment]
Rebased from trunk to fix conflicts.

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java (hunk 3 above)
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/network/Selector.java]: This code change has been reverted from this code review. Resolving the comment.

**Comment by mimaison** [comment]
@divijvaidya Can you rebase on trunk to resolve conflicts?
@ijuma Do you have any further comments or is this good to merge?

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java [hunk 8]
```diff
@@ -679,12 +678,13 @@ public static MemoryRecords withRecords(byte magic, long initialOffset, Compress
         long logAppendTime = RecordBatch.NO_TIMESTAMP;
         if (timestampType == TimestampType.LOG_APPEND_TIME)
             logAppendTime = System.currentTimeMillis();
-        MemoryRecordsBuilder builder = new MemoryRecordsBuilder(bufferStream, magic, compressionType, timestampType,
+        try (final MemoryRecordsBuilder builder = new MemoryRecordsBuilder(bufferStream, magic, compressionType, timestampType,
```
Comment (divijvaidya) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:679]: Note to reviewer(s):

`builder.build()` calls the `MemoryRecordsBuilder#close()` method, hence, this is not a leak in happy cases. But it will not call `close()` in case exception occurs at `builder.append()`, hence, MemoryRecordsBuilder will not be closed when an exception is thrown.

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java [hunk 9]
```diff
@@ -196,25 +196,24 @@ private static FilterResult filterTo(TopicPartition partition, Iterable<MutableR
...
                         deleteHorizonMs = batch.deleteHorizonMs().orElse(RecordBatch.NO_TIMESTAMP);
-                    builder = buildRetainedRecordsInto(batch, retainedRecords, bufferOutputStream, deleteHorizonMs);
-
-                    MemoryRecords records = builder.build();
-                    int filteredBatchSize = records.sizeInBytes();
-                    if (filteredBatchSize > batch.sizeInBytes() && filteredBatchSize > maxRecordBatchSize)
-                        log.warn("Record batch from {} with last offset {} exceeded max record batch size {} after cleaning " +
-                                        "(new size is {}). Consumers with version earlier than 0.10.1.0 may need to " +
-                                        "increase their fetch sizes.",
+                    try (final MemoryRecordsBuilder builder = buildRetainedRecordsInto(batch, retainedRecords, bufferOutputStream, deleteHorizonMs)) {
```
Comment (divijvaidya) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:202]: Note to reviewer(s):

`builder.build()` (in the next line) calls the MemoryRecordsBuilder#close() method, hence, this is not a leak here but this part of changes for code robustness (in case we start throwing exceptions from `buildRetainedRecordsInto` in future)

File: clients/src/main/java/org/apache/kafka/common/record/LegacyRecord.java [hunk 10]
```diff
@@ -385,8 +385,7 @@ private static void write(ByteBuffer buffer,
                               ByteBuffer value,
                               CompressionType compressionType,
                               TimestampType timestampType) {
-        try {
-            DataOutputStream out = new DataOutputStream(new ByteBufferOutputStream(buffer));
+        try (DataOutputStream out = new DataOutputStream(new ByteBufferOutputStream(buffer))) {
```
Comment (divijvaidya) [inline clients/src/main/java/org/apache/kafka/common/record/LegacyRecord.java:388]: Note to reviewer(s):

This is a NoOp since `close()` method of `ByteBufferOutputStream` doesn't do anything. This change is for code hygiene.

**Comment by divijvaidya** [comment]
@mimaison I have rebased from trunk ✅

**Comment by mimaison** [comment]
@divijvaidya You had 1 approval, so if you rebase you could merge this

File: gradle/dependencies.gradle [hunk 11]
```diff
@@ -165,7 +165,7 @@ versions += [
   spotbugs: "4.8.0",
   zinc: "1.9.2",
   zookeeper: "3.8.3",
-  zstd: "1.5.5-6"
```
Comment (mimaison) [inline gradle/dependencies.gradle]: We shouldn't bump dependencies in this PR. Let's have a different PR if we want to bump zstd.

Comment (divijvaidya) [reply gradle/dependencies.gradle]: Ah! of course. This change missed my attention.

**Review by mimaison — APPROVED** [review]
LGTM, thanks

**Comment by divijvaidya** [comment]
Unrelated test failures since they are known to be flaky as per https://ge.apache.org/scans/tests?search.rootProjectNames=kafka&search.tags=trunk&search.timeZoneId=Europe%2FBerlin 
```
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testReplicateFromLatest()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.connect.mirror.integration/IdentityReplicationIntegrationTest/Build___JDK_17_and_Scala_2_13___testReplicateFromLatest__/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest.testReplicateFromLatest()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationBaseTest/Build___JDK_17_and_Scala_2_13___testReplicateFromLatest__/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationExactlyOnceTest/Build___JDK_17_and_Scala_2_13___testOffsetTranslationBehindReplicationFlow__/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest.testReplicateFromLatest()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationSSLTest/Build___JDK_17_and_Scala_2_13___testReplicateFromLatest__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationTransactionsTest.testReplicateFromLatest()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationTransactionsTest/Build___JDK_11_and_Scala_2_13___testReplicateFromLatest__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationTransactionsTest.testReplicateFromLatest()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationTransactionsTest/Build___JDK_11_and_Scala_2_13___testReplicateFromLatest___2/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testReplicateFromLatest()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsWithCustomForwardingAdminIntegrationTest/Build___JDK_11_and_Scala_2_13___testReplicateFromLatest__/)
[Build / JDK 11 and Scala 2.13 / kafka.api.PlaintextConsumerTest.testShrinkingTopicSubscriptions(String, String).quorum=kraft+kip848.groupProtocol=consumer](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/kafka.api/PlaintextConsumerTest/Build___JDK_11_and_Scala_2_13___testShrinkingTopicSubscriptions_String__String__quorum_kraft_kip848_groupProtocol_consumer/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.TopicCommandIntegrationTest.testDescribeUnderReplicatedPartitionsWhenReassignmentIsInProgress(String).kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.tools/TopicCommandIntegrationTest/Build___JDK_11_and_Scala_2_13___testDescribeUnderReplicatedPartitionsWhenReassignmentIsInProgress_String__kraft/)
[Build / JDK 8 and Scala 2.12 / kafka.api.SaslSslConsumerTest.testCoordinatorFailover(String, String).quorum=zk.groupProtocol=generic](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/kafka.api/SaslSslConsumerTest/Build___JDK_8_and_Scala_2_12___testCoordinatorFailover_String__String__quorum_zk_groupProtocol_generic/)
[Build / JDK 8 and Scala 2.12 / kafka.network.ConnectionQuotasTest.testListenerConnectionRateLimitWhenActualRateAboveLimit()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/kafka.network/ConnectionQuotasTest/Build___JDK_8_and_Scala_2_12___testListenerConnectionRateLimitWhenActualRateAboveLimit__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.ConsistencyVectorIntegrationTest.shouldHaveSamePositionBoundActiveAndStandBy](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.streams.integration/ConsistencyVectorIntegrationTest/Build___JDK_8_and_Scala_2_12___shouldHaveSamePositionBoundActiveAndStandBy/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.StoreQueryIntegrationTest.shouldQuerySpecificStalePartitionStores()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.streams.integration/StoreQueryIntegrationTest/Build___JDK_8_and_Scala_2_12___shouldQuerySpecificStalePartitionStores__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.tools.MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful [2] Type=Raft-Isolated, MetadataVersion=3.8-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12228/12/testReport/junit/org.apache.kafka.tools/MetadataQuorumCommandTest/Build___JDK_8_and_Scala_2_12___testDescribeQuorumStatusSuccessful__2__Type_Raft_Isolated__MetadataVersion_3_8_IV0__Security_PLAINTEXT/)
```
